src/compiler/checker.ts

```diff
@@ -23842,7 +23842,9 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             let matched = false;
             for (let i = 0; i < types.length; i++) {
                 if (include[i]) {
-                    const targetType = getTypeOfPropertyOrIndexSignatureOfType(types[i], propertyName);
+                    const targetType = typeof propertyName === "number"
+                        ? getContextualTypeForElementExpression(types[i], propertyName)
+                        : getTypeOfPropertyOrIndexSignatureOfType(types[i], propertyName);

```

Comment:
found that this was not working like I wanted with variadic tuple types, so for tuples im using `getContextualTypeForElementExpression`.  Not sure if there's a better way to do this, also considered making this function an argument